src/water_cmd.cpp

```diff
@@ -452,10 +454,15 @@ CommandCost CmdBuildCanal(TileIndex tile, DoCommandFlag flags, uint32 p1, uint32
 	TileArea ta(tile, p1);
 
 	/* Outside the editor you can only drag canals, and not areas */
-	if (_game_mode != GM_EDITOR && ta.w != 1 && ta.h != 1) return CMD_ERROR;
+	if (_game_mode != GM_EDITOR) {
+		TileArea ta(tile, p1);
+		if (ta.w != 1 && ta.h != 1) return CMD_ERROR;
+	}
 
 	CommandCost cost(EXPENSES_CONSTRUCTION);
-	TILE_AREA_LOOP(tile, ta) {
+	std::unique_ptr<TileIterator> iter(HasBit(p2, 2) ? (TileIterator *)new DiagonalTileIterator(tile, p1) : new OrthogonalTileIterator(tile, p1));
```

Comment:
Actually, why is that cast necessary at all? I wouldn't expect it to be

Also consider using std::make_unique, likely cleaner

Reply:
Without the cast, I get an `error C2446: ':': no conversion from 'OrthogonalTileIterator *' to 'DiagonalTileIterator *'`.

Reply:
They don't have a comment parent type ?

Reply:
I think the issue is the ternary requiring the same type on both sides. Convert it to a proper if/else, should work (and use std::make_unique :) )